### ISSUE TITLE
Feature request: Optional chaining operator

### ISSUE THREAD

**Issue by Rosuav** [jquery/esprima#2112]
### Steps to reproduce

```js
esprima.parse('foo?.bar')
```

### Expected output

Some indication that this is optional chaining (if foo exists, it's the same as foo.bar, else it's undefined)

### Actual output

"Unexpected token ."

### Relevant references

https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#prod-OptionalExpression

This is supported in all major browsers and Node.js.

**Comment by jmelvin** [comment]
We started using more ES6 syntax recently and also ran into the optional chaining failure.  It guess we will have to be more selective until this is supported.

**Comment by sergey-v9** [comment]
we have had to move to [nightly-esprima](https://www.npmjs.com/package/nightly-esprima) because of this

**Comment by mariocampacci** [comment]
it is [not implemented](https://github.com/jquery/esprima/blob/512cd66c6ffd6083144b0150f09670e426252776/src/parser.ts#L1470) (?) 
although it seems limited to template literal as suggested by the comment

however I cannot understend why: ES6 shouldn't be handled? 57dc451e6d5dc9d692697561e402d532c9f432d0
and as far as i know the [proposal](https://github.com/tc39/proposal-optional-chaining) [passed](https://github.com/tc39/proposals/blob/main/finished-proposals.md)

**Comment by jogibear9988** [comment]
you can also try my esprima fork esprima-next
wich is in es6 and should work with ES6 code
https://github.com/node-projects/esprima-next

**Comment by Rosuav** [comment]
Does that work with the Python bindings?

**Comment by jogibear9988** [comment]
don't know

**Comment by lmh-justdo** [comment]
Object deconstruction like ... not support

**Comment by Nefcanto** [comment]
> we have had to move to [nightly-esprima](https://www.npmjs.com/package/nightly-esprima) because of this

@sergey-v9 what is that `nightly-esprima`? I searched it and it takes me back to the `esprima`  repository itself.

**Comment by sergey-v9** [comment]
@Nefcanto yes, it is the same, but bult from the latest version of the code in this repo from 2021 year (while the _official_ last version was 4.0.1 from 2018 year)
